File: sd.py
# syntax  name                          stack-before                     stack-after
# ==================================================================================
# [0-9]   push integer literal          []                            -> [literal]
# [A-Z]   push symbol literal           []                            -> [literal]
# {.*}    subroutine (see below)        []                            -> [start-ip]
# a       add                           [a b]                         -> [a+b]
# c       call                          [name]                        -> []
# d       divide                        [a b]                         -> [a/b]
# e       exchange (swap)               [a b]                         -> [b a]
# f       define subroutine             [start-ip name]               -> []
# h       halt                          []                            -> []
# i       conditional call              [condition if-true if-false]  -> []
# j       jump to subroutine            [symbol-or-relative]          -> []
# k       conditional jump              [condition if-true if-false]  -> []
# m       multiply                      [a b]                         -> [a*b]
# q       pop (discard)                 [a]                           -> []
# r       append stack top to result    [a]                           -> [a]
# s       subtract                      [a b]                         -> [a-b]
# t       swap with register            [a]                           -> [r]
# w       duplicate                     [a]                           -> [a a]
# x       index into stack              [index]                       -> [value]
# y       set stack value               [index value]                 -> []
# z       push stack length             []                            -> [top-index]

def run(code):
  code = list(code)

  stack = list()
  subroutines = dict()
  instructionpointer = 0
  callstack = list()
  result = list()
  register = 0

  def push_val(val):
    nonlocal stack
    stack.append(val)

  def pop_val():
    nonlocal stack
    return stack.pop()

  def skip():
    nonlocal code, instructionpointer
    level = 1
    while level > 0:
      instructionpointer += 1
      try:
        if code[instructionpointer] == '{':
          level += 1
        elif code[instructionpointer] == '}':
          level -= 1
      except IndexError:
        instructionpointer = len(code)

  while True:
    try:
      instruction = code[instructionpointer]
    except IndexError:
      break

    # push integer literal [] -> [literal]
    if '0' <= instruction <= '9':
      val = 9 - (57 - ord(instruction))
      push_val(val)

    # push symbol literal [] -> [literal]
    elif 'A' <= instruction <= 'Z':
      push_val(instruction)

    # subroutine [] -> [start-ip]
    elif instruction == '{':
      push_val(instructionpointer)
      ip_pre = instructionpointer
      skip()

    # return [] -> []
    elif instruction == '}':
      try:
        instructionpointer = callstack.pop()
      except IndexError:
        break

    # define subroutine [start-ip name] -> []
    elif instruction == 'f':
      name = pop_val()
      start = pop_val()
      subroutines[name] = start

    # add [a b] -> [a+b]
    elif instruction == 'a':
      b = pop_val()
      a = pop_val()
      push_val(a + b)

    # subtract [a b] -> [a-b]
    elif instruction == 's':
      b = pop_val()
      a = pop_val()
      push_val(a - b)

    # multiply [a b] -> [a*b]
    elif instruction == 'm':
      b = pop_val()
      a = pop_val()
      push_val(a * b)

    # divide [a b] -> [a/b]
    elif instruction == 'd':
      b = pop_val()
      a = pop_val()
      push_val(a // b)

    # jump [symbol-or-relative] -> []
    elif instruction == 'j':
      target = pop_val()
      if isinstance(target, str):
        instructionpointer = subroutines[target]
      else:
        instructionpointer += target
        continue

    # call subroutine [symbol] -> []
    elif instruction == 'c':
      callstack.append(instructionpointer)
      name = pop_val()
      instructionpointer = subroutines[name]

    # conditional call [condition if-true if-false] -> []
    elif instruction == 'i':
      if_false = pop_val()
      if_true = pop_val()
      cond = pop_val()
      if cond == 0:
        branch = if_false
      else:
        branch = if_true
      callstack.append(instructionpointer)
      instructionpointer = subroutines[branch]

    # conditional jump [condition if-true if-false] -> []
    elif instruction == 'k':
      if_false = pop_val()
      if_true = pop_val()
      cond = pop_val()
      if cond == 0:
        branch = if_false
      else:
        branch = if_true
      if isinstance(branch, str):
        instructionpointer = subroutines[branch]
      else:
        instructionpointer += branch
        continue

    # append top to result [a] -> [a]
    elif instruction == 'r':
      result.append(stack[-1])

    # discard [a] -> []
    elif instruction == 'q':
      pop_val()

    # duplicate [a] -> [a a]
    elif instruction == 'w':
      push_val(stack[-1])

    # exchange [a b] -> [b a]
    elif instruction == 'e':
      a = pop_val()
      b = pop_val()
      push_val(a)
      push_val(b)

    # stack top [] -> [top-index]
    elif instruction == 'z':
      push_val(len(stack))

    # index into stack [index] -> [value]
    elif instruction == 'x':
      index = pop_val()
      push_val(stack[index])

    # set stack index [index value] -> []
    elif instruction == 'y':
      value = pop_val()
      index = pop_val()
      stack[index] = value

    # swap with register [a] -> [r]
    elif instruction == 't':
      val = pop_val()
      register, val = val, register
      push_val(val)

    # halt [] -> []
    elif instruction == 'h':
      break

    instructionpointer += 1

  return bytes(result)

if __name__ == '__main__':
  import sys
  try:
    program = sys.argv[1]
    with open(program, 'r') as f:
      program = f.read()
  except IndexError:
    program = sys.stdin.read()

  if program == '':
    print(f'Usage: {sys.argv[0]} program\n  or pass the program via stdin')
    sys.exit(1)

  print(run(program))

File: sd.js
var SD = { };

SD.TypeError = function SDTypeError(ip) {
  Error.call(this, 'SD: Type error at ip=' + ip);
}
SD.TypeError.prototype = Object.create(Error.prototype);
SD.TypeError.prototype.constructor = SD.TypeError;

SD.SubroutineError = function SDSubroutineError(ip) {
  Error.call(this, 'SD: Unknown subroutine at ip=' + ip);
}
SD.SubroutineError.prototype = Object.create(Error.prototype);
SD.SubroutineError.prototype.constructor = SD.SubroutineError;

SD.StateError = function SDStateError(ip, message) {
  Error.call(this, 'SD: State error at ip=' + ip + ': ' + message);
}
SD.StateError.prototype = Object.create(Error.prototype);
SD.StateError.prototype.constructor = SD.StateError;

SD.run = function run(code) {
  "use strict";
  code = code.split('');

  var stop = new Object();

  var st = [ ],
      sr = { },
      ip = 0,
      cs = [ ],
      rt = [ ],
      rg = 0;

  var instr;

  function skip() {
    var level = 1;
    while (level > 0) {
      ip += 1;
      instr = code[ip];
      if (instr === undefined) {
        return false;
      }
      if (instr === '{') {
        level += 1;
      } else if (instr === '}') {
        level -= 1;
      }
    }
    return true;
  }

  il: for (;;) {
    instr = code[ip];
    if (instr === undefined) {
      break il;
    }

    if ('0' <= instr && instr <= '9') { // [] -> [literal]
      var val = 9 - (57 - instr.charCodeAt(0));
      st.push(val);
    } else if ('A' <= instr && instr <= 'Z') { // [] -> [literal]
      st.push(instr);
    } else if (instr === '{') { // [] -> [start-ip]
      st.push({ ip: ip });
      if ( ! skip()) {
        break il;
      }
    } else if (instr === '}') { // [] -> []
      ip = cs.pop();
      if (ip === undefined) {
        break il;
      }
    } else if (instr === 'f') { // [start-ip name] -> []
      var name = st.pop(),
          start = st.pop();
      if (typeof name !== 'string' ||
          typeof start !== 'object' ||
          ! ('ip' in start)) {
        throw new SD.TypeError(ip);
      }
      sr[name] = start;
    } else if (instr === 'a' || instr === 's' || instr === 'm' || instr === 'd') { // [a b] -> [c]
      var b = st.pop(),
          a = st.pop();
      var val;
      switch (instr) {
        case 'a': val = a + b; break;
        case 's': val = a - b; break;
        case 'm': val = a * b; break;
        case 'd': val = Math.floor(a / b); break;
      }
      st.push(val);
    } else if (instr === 'j') { // [symbol-or-relative] -> []
      var target = st.pop();
      if (typeof target === 'number') {
        ip += target;
        continue il;
      } else if (typeof target === 'string') {
        target = sr[target];
        if (target === undefined) {
          throw new SD.SubroutineError(ip);
        }
        ip = target.ip;
      } else {
        throw new SD.TypeError(ip);
      }
    } else if (instr === 'c') { // [symbol] -> []
      var target = st.pop();
      if (typeof target !== 'string') {
        throw new SD.TypeError(ip);
      }
      target = sr[target];
      if (target === undefined) {
        throw new SD.SubroutineError(ip);
      }
      cs.push(ip);
      ip = target.ip;
    } else if (instr === 'i') { // [condition true false] -> []
      var branchFalse = st.pop(),
          branchTrue = st.pop(),
          condition = st.pop();
      if (typeof condition !== 'number') {
        throw new SD.TypeError(ip);
      }
      var branch = condition === 0 ? branchFalse : branchTrue;
      if (typeof branch !== 'string') {
        throw new SD.TypeError(ip);
      }
      var target = sr[branch];
      if (target === undefined) {
        throw new SD.SubroutineError(ip);
      }
      cs.push(ip);
      ip = target.ip;
    } else if (instr === 'k') { // [condition true false] -> []
      var branchFalse = st.pop(),
          branchTrue = st.pop(),
          condition = st.pop();
      if (typeof condition !== 'number') {
        throw new SD.TypeError(ip);
      }
      var branch = condition === 0 ? branchFalse : branchTrue;
      if (typeof branch === 'number') {
        ip += branch;
        continue il;
      } else if (typeof branch === 'string') {
        var target = sr[branch];
        if (target === undefined) {
          throw new SD.SubroutineError(ip);
        }
        ip = target.ip;
      } else {
        throw new SD.TypeError(ip);
      }
    } else if (instr === 'r') { // [a] -> [a]
      var val = st[st.length - 1];
      if (typeof val !== 'number') {
        throw new SD.TypeError(ip);
      }
      rt.push(val);
    } else if (instr === 'q') { // [a] -> []
      st.pop();
    } else if (instr === 'w') { // [a] -> [a a]
      st.push(st[st.length - 1]);
    } else if (instr === 'e') { // [a b] -> [b a]
      var a = st.pop(),
          b = st.pop();
      st.push(a, b);
    } else if (instr === 'z') { // [] -> [top-index]
      st.push(st.length);
    } else if (instr === 'x') { // [index] -> [value]
      var index = st.pop();
      if (typeof index !== 'number') {
        throw new SD.TypeError(ip);
      } else if (index > st.length) {
        throw new SD.StateError(ip, 'index higher than stack top');
      }
      st.push(st[index]);
    } else if (instr === 'y') { // [index value] -> []
      var value = st.pop(),
          index = st.pop();
      if (typeof index !== 'number') {
        throw new SD.TypeError(ip);
      } else if (index > st.length) {
        throw new SD.StateError(ip, 'index higher than stack top');
      }
      st[index] = value;
    } else if (instr === 't') { // [a] -> [r]
      var value = st.pop();
      if (typeof value !== 'number') {
        throw new SD.TypeError(ip);
      }
      st.push(rg);
      rg = value;
    } else if (instr === 'h') { // [] -> []
      break il;
    }

    ip += 1;
  }

  return rt;
}

if (global) {
  global.SD = SD;
}
